### PR TITLE
Updates Client to use async/await

### PR DESCRIPTION
**Pull request by douglance**
- Improves error handling
- Deprecates the callback API
- Bumps to `v2.0.0` due to breaking API

### CODE REVIEW

File: src/client.ts [hunk 1]
```diff
@@ -1365,7 +990,7 @@ export class Client {
     while (off + 4 < decResLengths.getAddresses) {
       if (arePubkeys) {
         // Pubkeys are shorter and are returned as buffers
-        const pubBytes = addrData.slice(off, off + 65);
+        const pubBytes = decrypted.slice(off, off + 65);
```
[the remark this answers is not in the document]
Comment (douglance) [reply src/client.ts:994]: `_handleEncResponse()` used to return `{ err: any, data: any }`, but now it just returns  `any`. This is because with async/await, it's a lot easier to do error handling. So we don't have to pass around `{err, data}` objects, we can just pass around the data and throw errors if things go wrong.

File: src/client.ts [hunk 2]
```diff
@@ -1398,19 +1024,15 @@ export class Client {
...
+    this._parseAndAddWallets(decrypted.slice(65));
     if (
       this.activeWallets.internal.uid.equals(EMPTY_WALLET_UID) &&
       this.activeWallets.external.uid.equals(EMPTY_WALLET_UID)
     ) {
-      return 'No active wallet.';
+      throw new Error('No active wallet.');
     }
-    return null;
+    return decrypted;
```
[the remark this answers is not in the document]
Comment (douglance) [reply src/client.ts:1036]: I just looked, and I don't think anyone is actually using `fetchActiveWallets()` which uses this function.

File: src/client.ts [hunk 3]
```diff
@@ -322,578 +287,277 @@ export class Client {
...
-        flags.indexOf(flag) > -1 &&
-        (flag === EXTERNAL.GET_ADDR_FLAGS.ED25519_PUB ||
-          flag === EXTERNAL.GET_ADDR_FLAGS.SECP256K1_PUB);
-      if (!isPubkeyOnly && !isValidAssetPath(startPath, fwConstants)) {
-        return cb('Parent derivation path is not supported');
-      }
+  public async getAddresses (opts: { startPath: number[], n: number, flag?: number }): Promise<Buffer[]> {
+    const MAX_ADDR = 10;
+    const { startPath, n, flag = 0 } = opts;
+    if (startPath === undefined || n === undefined)
```
[the remark this answers is not in the document]
Comment (douglance) [reply src/client.ts:293]: 5785e66ab4f0776bc18211c634a7f9b05b9329e4

File: src/client.ts [hunk 4]
```diff
@@ -322,578 +287,277 @@ export class Client {
...
-      off++;
-      return this._request(payload, 'GET_ADDRESSES', (err, res) => {
-        if (err) return cb(err);
-        const parsedRes = this._handleGetAddresses(res, flagVal);
-        if (parsedRes.err) return cb(parsedRes.err);
-        return cb(null, parsedRes.data);
-      });
-    });
+    const fwConstants = getFwVersionConst(this.fwVersion);
+    const flags = fwConstants.getAddressFlags || [];
```
[the remark this answers is not in the document]
Comment (douglance) [reply src/client.ts]: 5785e66ab4f0776bc18211c634a7f9b05b9329e4

File: src/client.ts [hunk 5]
```diff
@@ -322,578 +287,277 @@ export class Client {
...
+        throw new Error('Transaction is too large');
+      reqPayload = req.payload;
+      schema = req.schema;
+    }
+    // Build the payload
+    const payload = Buffer.alloc(2 + fwConstants.reqMaxDataSz);
+    let off = 0;
+    // Whether there will be follow up requests
+    const hasExtraPayloads =
+      req.extraDataPayloads && Number(req.extraDataPayloads.length > 0);
```
[the remark this answers is not in the document]
Comment (douglance) [reply src/client.ts:456]: I think it's supposed to be that way so it can write that value to the buffer.

File: src/client.ts [hunk 6]
```diff
@@ -902,153 +566,125 @@ export class Client {
...
+      throw new Error(
+        'One or more key-value mapping must be provided in `records` param.',
+      );
+    } else if (Object.keys(records).length > fwConstants.kvActionMaxNum) {
+      throw new Error(
+        `Too many keys provided. Please only provide up to ${fwConstants.kvActionMaxNum}.`,
+      );
+    } else if (Object.keys(records).length < 1) {
+      throw new Error('You must provide at least one key to add.');
+    }
```
[the remark this answers is not in the document]
Comment (douglance) [reply src/client.ts:590]: I don't want to touch it because it uses that check to write something to the buffer.

File: src/client.ts [hunk 7]
```diff
@@ -251,69 +235,50 @@ export class Client {
...
+    if (pairingSecret.length > 0) {
+      // If a pairing secret of zero length is passed in, it usually indicates we want to cancel
+      // the pairing attempt. In this case we pass a zero-length name buffer so the firmware can
+      // know not to draw the error screen. Note that we still expect an error to come back
+      // (RESP_ERR_PAIR_FAIL)
+      nameBuf.write(this.name);
+    }
+    const hash = generateAppSecret(pubKey, nameBuf, Buffer.from(pairingSecret));
+    const sig = this.key.sign(hash); // returns an array, not a buffer
```
[the remark this answers is not in the document]
Comment (douglance) [reply src/client.ts:253]: I believe so. I haven't touched that. It's from the `elliptic` lib.

File: src/client.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (alex-miller-0) [reply src/client.ts:456]: Oh you're doing a boolean check. Oops.

File: src/client.ts [hunk 8]
```diff
@@ -322,578 +287,277 @@ export class Client {
...
+
+    // WalletUID
+    const wallet = this.getActiveWallet();
+    if (wallet === null) {
+      throw new Error('No active wallet.');
+    }
+    wallet.uid.copy(payload, off);
+    off += 32;
+    // Build the start path (5x u32 indices)
+    if (fwConstants.varAddrPathSzAllowed) {
```
[the remark this answers is not in the document]
Comment (alex-miller-0) [reply src/client.ts:332]: Looks like you removed it or else I misread originally. Either way looks good now

File: src/client.ts [hunk 9]
```diff
@@ -1300,28 +934,24 @@ export class Client {
...
-    try {
-      this.ephemeralPub = getP256KeyPairFromPub(pub);
-      return { err: null, data: res };
-    } catch (e) {
-      return {
-        err: `Error handling getAddresses response: ${e.toString()}`,
-      };
-    }
+    this.ephemeralPub = getP256KeyPairFromPub(pub);
+    return res
```
Comment (douglance) [inline src/client.ts:954]: @alex-miller-0 Here's where the return value of `_handleEncResponse` was changed.

Comment (alex-miller-0) [reply src/client.ts:954]: Oh ty. Missed it in the diff

**Comment by douglance** [comment]
## test

```text
  Connect and Pair
    ✔ Should connect to a Lattice
    ✔ Should attempt to pair with pairing secret
    ✔ Should try to connect again but recognize the pairing already exists (1166ms)
    ✔ Should test SDK dehydration/rehydration (5286ms)
    ✔ Should get addresses (14983ms)
Please REJECT the next request if the warning screen displays. Press enter to continue.
    ✔ Should sign Ethereum transactions (13073ms)
    ✔ Should sign legacy Bitcoin inputs (3658ms)
    ✔ Should sign wrapped segwit Bitcoin inputs (3661ms)
    ✔ Should sign wrapped segwit Bitcoin inputs to a bech32 address (3666ms)
    ✔ Should sign an input from a native segwit account (3641ms)


  10 passing (49s)
```

## test-btc

```text
  Fetch wallet seed
    ✔ Should connect to a Lattice and make sure it is already paired. (1164ms)

  exportSeed
    ✔ Should get GP_SUCCESS for a known, connected wallet (2695ms)

  Test segwit spender (p2wpkh)
    ✔ Should test p2wpkh->p2pkh (33102ms)
    ✔ Should test p2wpkh->p2sh-p2wpkh (35604ms)
    ✔ Should test p2wpkh->p2wpkh (32847ms)

  Test wrapped segwit spender (p2sh-p2wpkh)
    ✔ Should test p2sh-p2wpkh->p2pkh (24958ms)
    ✔ Should test p2sh-p2wpkh->p2sh-p2wpkh (31983ms)
    ✔ Should test p2sh-p2wpkh->p2wpkh (27476ms)

  Test legacy spender (p2pkh)
    ✔ Should test p2pkh->p2pkh (25018ms)
    ✔ Should test p2pkh->p2sh-p2wpkh (27423ms)
    ✔ Should test p2pkh->p2wpkh (24939ms)


  11 passing (4m)
```

## test-eth-msg

 ```text
  Setup client
    ✔ Should setup the test client (64ms)
    ✔ Should connect to a Lattice and make sure it is already paired. (1257ms)

  Test ETH personalSign
    ✔ Should throw error when message contains non-ASCII characters
    ✔ Should test ASCII buffers (7357ms)
    ✔ Should test hex buffers (3124ms)
    ✔ Should test a message that needs to be prehashed (3224ms)
    ✔ Msg: sign_personal boundary conditions and auto-rejected requests (8498ms)
    ✔ Should test 20 random payloads (82943ms)

  Test ETH EIP712
    ✔ Should test a message that needs to be prehashed (3144ms)
    ✔ Should test simple dydx example (3171ms)
    ✔ Should test a Loopring message with non-standard numerical type (3719ms)
    ✔ Should test a large 1inch transaction (5868ms)
    ✔ Should test an example with 0 values (3123ms)
    ✔ Should test canonical EIP712 example (3207ms)
    ✔ Should test canonical EIP712 example with 2nd level nesting (3683ms)
    ✔ Should test canonical EIP712 example with 3rd level nesting (3703ms)
    ✔ Should test canonical EIP712 example with 3rd level nesting and params in a different order (3665ms)
    ✔ Should test a payload with an array type (3634ms)
    ✔ Should test multiple array types (3727ms)
    ✔ Should test a nested array (3150ms)
    ✔ Should test a nested array of custom type (3656ms)
    ✔ Should test a bunch of EIP712 data types (4141ms)
    ✔ Should test a payload with a nested type in multiple nesting levels (3177ms)
    ✔ Should test a payload that requires use of extraData frames (6419ms)
    ✔ Should test a message with very large types (5868ms)
    ✔ Should test random edge case #1 (3149ms)
    ✔ Should test 20 random payloads (73037ms)


  27 passing (4m)
  ```

## test-kv

```text
  Test key-value files API
    ✔ Should connect to a Lattice
    ✔ Should attempt to pair with pairing secret
    ✔ Should try to connect again but recognize the pairing already exists (1189ms)
Do you want to clear all kv records and start anew? (Y/N) y
    ✔ Should ask if the user wants to reset state (21388ms)
Please reject if you do NOT see an address tag.
Using the legacy ETH signing path. This will soon be deprecated. Please switch to general signing request.
    ✔ Should make a request to an unknown address (7009ms)
    ✔ Should get the initial set of records (1657ms)
    ✔ Should add some key value records (6159ms)
    ✔ Should fail to add records with unicode characters (8336ms)
    ✔ Should fail to add zero length keys and values
    ✔ Should fetch the newly created records (1662ms)
Please reject if you do NOT see an address tag.
Using the legacy ETH signing path. This will soon be deprecated. Please switch to general signing request.
    ✔ Should make a request to an address which is now known (12045ms)
Please reject if you do NOT see an address tag.
    ✔ Should make an EIP712 request that uses the record (9811ms)
Please reject if you do NOT see an address tag.
Using the legacy ETH signing path. This will soon be deprecated. Please switch to general signing request.
    ✔ Should make a request with calldata (13241ms)
    ✔ Should remove key value records (1664ms)
    ✔ Should confirm the records we recently added are removed (2141ms)
    ✔ Should add the same record with case sensitivity (4656ms)
Please reject if you do NOT see an address tag.
Using the legacy ETH signing path. This will soon be deprecated. Please switch to general signing request.
    ✔ Should make another request to make sure case sensitivity is enforced (43078ms)
    ✔ Should get the id of the newly added record (1649ms)
    ✔ Should remove the new record (2136ms)
    ✔ Should confirm there are no new records (2149ms)


  20 passing (2m)
  ```

## test-non-exportable

TODO: Have to reflash firmware

## test-signing

```text
  Test General Signing
    ✔ Should connect to a Lattice and make sure it is already paired. (1150ms)
    ✔ Should export the seed (2754ms)
    ✔ Should load determinism tests (296ms)
    ✔ Should load unformatted tests
    ✔ Should load Solana tests (88ms)
    ✔ Should load Terra tests (281ms)
    ✔ Should load EVM tests (271ms)

  [Determinism]
    Setup and validate seed
      ✔ Should remove the seed (5295ms)
      ✔ Should load the known test seed (3151ms)

Please remove, re-insert, and unlock your SafeCard.
Press enter to continue after addresses have fully synced.
      ✔ Should wait for the user to remove and re-insert the card (triggering SafeCard wallet sync) (15180ms)
      ✔ Should re-connect to the Lattice and update the walletUID. (1227ms)
      ✔ Should ensure export seed matches the test seed (2698ms)
      ✔ Should validate some Ledger addresses derived from the test seed (6893ms)
Please switch to a SafeCard with a different seed. Press enter to continue.
Please switch back to the first SafeCard. Press enter to continue.
      ✔ Should test that wrongWallet retry works (41023ms)
    Test uniformity of Ethereum transaction sigs
      ✔ Should validate uniformity sigs on m/44'/60'/0'/0/0 (19391ms)
      ✔ Should validate uniformity sigs on m/44'/60'/0'/0/1 (16785ms)
      ✔ Should validate uniformity sigs on m/44'/60'/0'/0/8 (16284ms)
      ✔ Should update the transaction to use oversized data
      ✔ Should validate uniformity sigs on m/44'/60'/0'/0/0 (15800ms)
      ✔ Should validate uniformity sigs on m/44'/60'/0'/0/1 (16345ms)
      ✔ Should validate uniformity sigs on m/44'/60'/0'/0/8 (18245ms)
    Compare personal_sign signatures vs Ledger vectors (1)
      ✔ Should validate signature from addr0 (3144ms)
      ✔ Should validate signature from addr1 (3140ms)
      ✔ Should validate signature from addr8 (3135ms)
    Compare personal_sign signatures vs Ledger vectors (2)
      ✔ Should validate signature from addr0 (3197ms)
      ✔ Should validate signature from addr1 (3220ms)
      ✔ Should validate signature from addr8 (3139ms)
    Compare personal_sign signatures vs Ledger vectors (3)
      ✔ Should validate signature from addr0 (13669ms)
      ✔ Should validate signature from addr1 (7199ms)
      ✔ Should validate signature from addr8 (3136ms)
    Compare EIP712 signatures vs Ledger vectors (1)
      ✔ Should validate signature from addr0 (3661ms)
      ✔ Should validate signature from addr1 (9132ms)
      ✔ Should validate signature from addr8 (9471ms)
    Compare EIP712 signatures vs Ledger vectors (2)
      ✔ Should validate signature from addr0 (3157ms)
      ✔ Should validate signature from addr1 (4666ms)
      ✔ Should validate signature from addr8 (3157ms)
    Compare EIP712 signatures vs Ledger vectors (3)
      ✔ Should validate signature from addr0 (3183ms)
      ✔ Should validate signature from addr1 (3151ms)
      ✔ Should validate signature from addr8 (3646ms)

  [Unformatted]
    ✔ Should test pre-hashed messages (18228ms)
    ✔ Should test ASCII text formatting (3156ms)
    ✔ Should validate SECP256K1/KECCAK signature against dervied key (6878ms)
    ✔ Should validate ED25519/NULL signature against dervied key (3184ms)
    ✔ Should validate SECP256K1/KECCAK signature against ETH_MSG request (legacy) (6302ms)
    ✔ Should test random payloads (60010ms)

  [Solana]
    ✔ Should test and validate Solana transaction encoding (7416ms)

  [Terra]
    ✔ Should test address derivations (10835ms)
    ✔ Should decode MsgSend (3633ms)
    ✔ Should decode MsgMultiSend (6895ms)
    ✔ Should decode MsgExecuteContract (3635ms)
    ✔ Should fail decode unsupported message type but still decode the tx (3671ms)
    ✔ Should sign a legacy Amino payload without the decoder (3673ms)

  [EVM] Test transactions
    EIP1559
      ✔ Should test a basic transaction (3134ms)
      ✔ Should test a Rinkeby transaction (3126ms)
      ✔ Should test a transaction with an access list (3671ms)
    EIP2930
      ✔ Should test a basic transaction (3162ms)
      ✔ Should test a Rinkeby transaction (3171ms)
      ✔ Should test a transaction with an access list (3161ms)
    Legacy (Non-EIP155)
      ✔ Should test a transaction that does not use EIP155 (3195ms)
    Boundary tests
      ✔ Should test shorter derivation paths (10642ms)
      ✔ Should test other chains (21138ms)
      ✔ Should test range of `value` (7790ms)
      ✔ Should test range of `data` size (12021ms)
      ✔ Should test contract deployment (3698ms)
      ✔ Should test direct RLP-encoded payoads with bad params (16886ms)
    Random Transactions
      3) Should test random transactions
    [TODO: deprecate] Test Legacy Pathway (while it still exists)
Using the legacy ETH signing path. This will soon be deprecated. Please switch to general signing request.
      ✔ Should test legacy signing for legacy EIP155 transaction (3650ms)
Using the legacy ETH signing path. This will soon be deprecated. Please switch to general signing request.
      ✔ Should test legacy signing for EIP1559 (3728ms)
Using the legacy ETH signing path. This will soon be deprecated. Please switch to general signing request.
      ✔ Should test a Polygon transaction (chainId=137) (3146ms)

  [EVM] Test decoders
    Test ABI decoder vectors
(node:25763) [DEP0066] DeprecationWarning: OutgoingMessage.prototype._headers is deprecated
(Use `node --trace-deprecation ...` to show where the warning was created)
      ✔ (Etherscan + 4byte #0) 0xeee0752109c6d31038bab6c2b0a3e3857e8bffb9c229de71f0196fda6fb28a5e (13817ms)
      ✔ (Etherscan + 4byte #1) 0xa6173b4890303e12ca1b195ea4a04d8891b0d83768b734d2ecdb9c6dd9d828c4 (7662ms)
      ✔ (Etherscan + 4byte #2) 0xee9710119c13dba6fe2de240ef1e24a2489e98d9c7dd5881e2901056764ee234 (7181ms)
      ✔ (Etherscan + 4byte #3) 0xc33308ca105630b99a0c24ddde4f4506aa4115ec6f1a25f1138f3bd8cfa32b49 (7212ms)
      ✔ (Etherscan + 4byte #4) 0xe15e6205c6696d444fc426468defdf08e89a0f5b3b8a17c68428f7aeefd53ca1 (12434ms)
      ✔ (Canonical #0) swapExactTokensForTokens(uint256,uint256,address[],address,uint256) (3155ms)
      ✔ (Canonical #1) proveAndClaimWithResolver((bytes,bytes)[]) (3144ms)
      ✔ (Canonical #2) proveAndClaimWithResolver(bytes,(bytes,bytes)[]) (3115ms)
      ✔ (Canonical #3) proveAndClaimWithResolver(bytes,(bytes,bytes)[],bytes,address,address) (3211ms)
      ✔ (Canonical #4) remove_liquidity_one_coin(uint256,int128,uint256) (9160ms)
      ✔ (Canonical #5) multidimArr(uint256[][],bool) (3658ms)
      ✔ (Canonical #6) multidimArr(uint256[2][]) (3651ms)
      ✔ (Canonical #7) multidimArr(uint256[][2]) (3163ms)
      ✔ (Canonical #8) multidimArr(uint256[2],bool) (3145ms)
      ✔ (Canonical #9) multidimArr(uint256[2][2],bool) (4736ms)
      ✔ (Canonical #10) multidimArr(uint256[2][2][2],bool) (3254ms)
      ✔ (Canonical #11) multidimArr(uint256[2][2],bool) (3160ms)
      ✔ (Canonical #12) multidimArr(uint256[][][],bool) (3138ms)
      ✔ (Canonical #13) multidimArr(bool,uint256[],bool) (3668ms)
      ✔ (Canonical #14) multidimArr(bytes,uint256[][],bool) (3718ms)
      ✔ (Canonical #15) singleTup((uint256,uint256)) (3157ms)
      ✔ (Canonical #16) singleTup((uint256[])) (3161ms)
      ✔ (Canonical #17) singleTup((uint256[],bool)) (3152ms)
      ✔ (Canonical #18) singleTup((bool,bool,uint[],uint256)) (3135ms)
      ✔ (Canonical #19) singleTup((uint256[1],uint256[])) (3159ms)
      ✔ (Canonical #20) singleTup((bytes,uint256)) (3152ms)
      ✔ (Canonical #21) singleTup((bytes,uint256)[]) (3148ms)
      ✔ (Canonical #22) singleTup((uint256,uint256)[]) (3628ms)
      ✔ (Canonical #23) singleTup((uint256,uint256)[2]) (3690ms)
      ✔ (Canonical #24) singleTup((uint256[2],uint256)) (3150ms)
      ✔ (Canonical #25) singleTup((uint256,uint256)[2]) (3671ms)
      ✔ (Canonical #26) singleTup((uint256,uint256)[],bytes) (3311ms)
      ✔ (Canonical #27) singleTup((uint256,uint256)[][],bytes) (3143ms)
      ✔ (Canonical #28) singleTup((uint256,uint256)[2][],bytes) (3657ms)
      ✔ (Canonical #29) singleTup((uint256,uint256)[][2],bytes) (3234ms)
      ✔ (Canonical #30) singleTup((uint256,uint256)[2],bytes) (4197ms)
      ✔ (Canonical #31) singleTup(bytes[],(uint256,uint256)[2],bytes) (3686ms)
      ✔ (Canonical #32) singleTup(bytes[2],(uint256,uint256)[2],bytes) (3178ms)
      ✔ (Canonical #33) singleTup((uint256,uint256)[2],bytes) (3273ms)
      ✔ (Canonical #34) singleTup((uint256,uint256)[2][2],bytes) (3184ms)
      ✔ (Canonical #35) singleTup((uint256,uint256)[2][2][2],bytes) (3166ms)
      ✔ (Canonical #36) singleTup((uint256),bool) (3170ms)
      ✔ (Canonical #37) singleTup(bytes,(bytes),bool) (3175ms)
      ✔ (Canonical #38) singleTup((uint256,bool),bool) (3137ms)
      ✔ (Canonical #39) singleTup((uint256,bool),bytes) (3118ms)
      ✔ (Canonical #40) singleTup(bytes,(uint256),bool) (3155ms)
      ✔ (Canonical #41) singleTup(bytes,(bytes),bool) (3149ms)
      ✔ (Canonical #42) singleTup(bool,(uint256,bytes),bytes) (3172ms)
      ✔ (Canonical #43) singleTup((uint256,uint256)[],bool) (3663ms)
      ✔ (Canonical #44) singleTup((bytes)[],bool) (3155ms)
      ✔ (Canonical #45) singleTup((bytes)[1],bool) (3176ms)
      ✔ (Canonical #46) multiTup((uint256,uint256)[],(bool)) (3179ms)
      ✔ (Canonical #47) multiTup((uint256,uint256)[2],(bool)) (3242ms)
      ✔ (Canonical #48) multiTup((uint256)[2],(bool)[]) (3151ms)
      ✔ (Canonical #49) multiTup((uint256)[2],(bool)[2]) (3153ms)
      ✔ (Canonical #50) multiTup((uint256)[],(bool)) (3184ms)
      ✔ (Canonical #51) multiTup((uint256)[],(bool)[2]) (3176ms)
      ✔ (Canonical #52) multiTup((uint256,uint256)[],(bool)[]) (3653ms)
      ✔ (Canonical #53) multiTup((uint256,uint256)[2],(bool)[]) (3134ms)
      ✔ (Canonical #54) multiTup((uint256)[2][2],(bool)[2][2]) (3656ms)
      ✔ (Canonical #55) multiTup((uint256)[2][],(bool)[2][]) (3217ms)
      ✔ (Canonical #56) multiTup((uint256)[2][],(bool)[][]) (3668ms)
      ✔ (Canonical #57) multiTup((uint256)[][],(bool)[2][]) (3151ms)
      ✔ (Canonical #58) multiTup((uint256)[][],(bool)[][]) (3143ms)
      ✔ (Canonical #59) multiTup((uint256)[][],(bool)[2][2]) (4662ms)
      ✔ (Canonical #60) multiTup((uint256)[2][2],(bool)[][]) (3684ms)
      ✔ (Canonical #61) nestedTup(uint256,(bytes)[1],bool) (3661ms)
      ✔ (Canonical #62) nestedTup((uint256,(bool,address),bool)) (3124ms)
      ✔ (Canonical #63) nestedTup((uint256,(bool,bytes),bool)) (3129ms)
      ✔ (Canonical #64) nestedTup(bool,(uint256,(bool,bytes),bool)) (3189ms)
      ✔ (Canonical #65) nestedTup(bytes,(uint256,(bool,bytes),bool)) (3173ms)
      ✔ (Canonical #66) nestedTup(bytes,(uint256,(bool,bytes)[],bool)) (3174ms)
      ✔ (Canonical #67) nestedTup(bool,(uint256,(bool,bytes),bool)[]) (3686ms)
      ✔ (Canonical #68) nestedTup(bytes,(uint256,(bool,bytes)[][],bool)[]) (5272ms)
      ✔ (Canonical #69) nestedTup(bytes,(uint256,(bool)[2],bool)) (3738ms)
      ✔ (Canonical #70) nestedTup((uint256,(bytes[])[2],uint256)) (3173ms)
      ✔ (Canonical #71) nestedTup((uint256,(bytes[2])[2],uint256)) (3173ms)
      ✔ (Canonical #72) nestedTup((uint256,(bytes[2])[],uint256)) (3144ms)
      ✔ (Canonical #73) nestedTup((uint256,(bytes[])[],uint256)) (3175ms)
      ✔ (Canonical #74) nestedTup((uint256,(bytes)[2],bool)) (3647ms)
      ✔ (Canonical #75) nestedTup((bytes,(bytes)[2],bool)) (3208ms)
      ✔ (Canonical #76) nestedTup(bytes,(uint256,(bool)[2],bool)) (3208ms)
      ✔ (Canonical #77) nestedTup(bytes,(uint256,(bytes)[2],bool)) (3130ms)
```

## test-wallet-jobs

```text
  Test Wallet Jobs
    ✔ Should connect to a Lattice and make sure it is already paired. (1168ms)

  exportSeed
    ✔ Should get GP_SUCCESS for a known, connected wallet (2678ms)
    ✔ Should get GP_ENODEV for unknown (random) wallet (2133ms)

  getAddresses
    ✔ Should get GP_SUCCESS for active wallet (2134ms)
    ✔ Should get GP_EWALLET for unknown (random) wallet (1677ms)
    ✔ Should get GP_EINVAL if `count` exceeds the max request size (2133ms)
    ✔ Should validate first ETH (2193ms)
    ✔ Should validate an ETH address from a different EVM coin type (2253ms)
    ✔ Should validate the first BTC address (2265ms)
    ✔ Should validate first BTC change address (2112ms)
    ✔ Should validate the first BTC address (testnet) (2129ms)
    ✔ Should validate first BTC change address (testnet) (2634ms)
    ✔ Should fetch a set of BTC addresses (2145ms)
    ✔ Should fetch a set of BTC addresses (bech32) (2130ms)
    ✔ Should fetch a set of BTC addresses (legacy) (2109ms)
    ✔ Should fetch address with nonstandard path (2220ms)
    ✔ Should fail to fetch from path with an unknown currency type
    ✔ Should validate address with pathDepth=4 (2103ms)
    ✔ Should validate address with pathDepth=3 (2131ms)
    ✔ Should validate random Bitcoin addresses of all types (19748ms)
    ✔ Should test export of SECP256K1 public keys (2127ms)
    ✔ Should test export of ED25519 public keys (4290ms)

  signTx
    ✔ Should get GP_SUCCESS for active wallet (2201ms)
    ✔ Should get GP_SUCCESS for signing out of shorter (but allowed) paths (2106ms)
    ✔ Should get GP_EWALLET for unknown (random) wallet (1596ms)
    ✔ Should get GP_EWALLET for known wallet that is inactive (2181ms)
    ✔ Should get GP_EINVAL when `numRequests` is 0 (1637ms)
    ✔ Should get GP_EINVAL when `numRequests` exceeds the max allowed (2160ms)
    ✔ Should return GP_EINVAL when a signer `pathDepth` is of invalid size (6961ms)
    ✔ Should get GP_SUCCESS when signing from a non-ETH EVM path (2756ms)

  Get delete permission

The following tests will remove your seed.
It should be added back in a later test, but these tests could fail!
Press enter to continue.
    ✔ Should get permission to remove seed. (17432ms)

  Test leading zeros
    ✔ Should remove the current seed (5138ms)
    ✔ Should load the new seed (3179ms)

Please remove your SafeCard, then re-insert and unlock it.
Press enter to continue.
    ✔ Should wait for the user to remove and re-insert the card (triggering SafeCard wallet sync) (25520ms)
    ✔ Should reconnect to update the wallet UIDs (1116ms)
    ✔ Should make sure the first address is correct (3158ms)
    ✔ Should test address m/44'/60'/0'/0/396 (1 leading zero byte) (7795ms)
    ✔ Should test address m/44'/60'/0'/0/406 (1 leading zero byte) (6298ms)
    ✔ Should test address m/44'/60'/0'/0/668 (1 leading zero byte) (5318ms)
    ✔ Should test address m/44'/60'/0'/0/71068 (2 leading zero bytes) (5329ms)
    ✔ Should test address m/44'/60'/0'/0/82173 (2 leading zero bytes) (5324ms)
    ✔ Should test address m/44'/60'/0'/0/11981831 (3 leading zero bytes) (5325ms)
    ✔ Should switch to testing public keys
    ✔ Should test address m/44'/60'/0'/153 (5323ms)
    ✔ Should prepare for one more derivation step
    ✔ Should test address m/44'/60'/0'/153/0 (7884ms)
    ✔ Should test address m/44'/60'/0'/153/1 (10852ms)
    ✔ Should test address m/44'/60'/0'/153/5 (5313ms)
    ✔ Should test address m/44'/60'/0'/153/10000 (5335ms)
    ✔ Should test address m/44'/60'/0'/153/9876543 (5326ms)

  deleteSeed
    ✔ Should get GP_EINVAL for unknown (random) wallet (1635ms)
    ✔ Should get GP_SUCCESS for a known, connected wallet. (5224ms)

  Load Original Seed Back
    ✔ Should get GP_EINVAL if `exportability` option is invalid (1621ms)
    ✔ Should get GP_SUCCESS when valid seed is provided to valid interface (3701ms)


Please remove your SafeCard, then re-insert and unlock it.
Press enter to continue.
    ✔ Should wait for the user to remove and re-insert the card (triggering SafeCard wallet sync) (34155ms)
    ✔ Should reconnect to update the wallet UIDs (1151ms)
    ✔ Should ensure export seed matches the seed we just loaded (2686ms)
    ✔ Should get GP_FAILURE if interface already has a seed (2137ms)
Please remove your SafeCard to run this test.
Press enter to continue.
    ✔ Should get GP_EAGAIN when trying to load seed into SafeCard when none exists (19364ms)

Please re-insert and unlock your SafeCard to continue.
Press enter to continue.
    ✔ Should wait for the card to be re-inserted (25993ms)
```

File: test/testKv.ts [hunk 10]
```diff
@@ -113,7 +113,7 @@ describe('Test key-value files API', () => {
       await client.sign(ETH_REQ);
     } catch (err) {
       expect(
-        err.indexOf(responseMsgs[responseCodes.RESP_ERR_USER_DECLINED])
+        err.message.indexOf(responseMsgs[responseCodes.RESP_ERR_USER_DECLINED])
```
[the remark this answers is not in the document]
Comment (douglance) [reply test/testKv.ts:116]: https://github.com/GridPlus/gridplus-sdk/pull/384#issuecomment-1145019751

File: src/client.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (douglance) [reply src/client.ts]: [a5a2382](https://github.com/GridPlus/gridplus-sdk/pull/384/commits/a5a2382873c276985ca5da490763a2aa8692d165)

File: src/client.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (douglance) [reply src/client.ts]: [a5a2382](https://github.com/GridPlus/gridplus-sdk/pull/384/commits/a5a2382873c276985ca5da490763a2aa8692d165)

**Review by alex-miller-0 — APPROVED** [review]
Wooo! Awesome work. Ready to merge pending rebase 👍
